feat(logs): make the addition of permissions to Lambda functions optional

zARODz11z opened the pull request:
### What does this PR do?
Makes the addition of permissions to Lambda functions optional for users instantiating the LambdaDestination class. By default the permissions will be added in hopes of not breaking existing user's configurations. Added a unit test which tests the scenario in which a user would like to opt out of adding permissions to their function.

### Motivation
Fixes #14198 in which users with a high amount of lambdas will cause `The final policy size (20497) is bigger than the limit (20480)` in our [Datadog CDK Construct Library](https://github.com/DataDog/datadog-cdk-constructs)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 401b558aad04dc4672fc6f3012a066cb07152850
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=07b6a49f-0af3-4da0-8d23-886376187182%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

zARODz11z commented:
Which README does the PR linter action want me to edit? The README for the aws-logs-destinations module is empty.

zARODz11z commented:
@rix0rrr @nija-at Hey let me know if there is anything else I could do on my end.

rix0rrr commented on packages/@aws-cdk/aws-logs-destinations/lib/lambda.ts:
```diff
@@ -7,19 +7,20 @@ import { Construct } from '@aws-cdk/core';
  */
 export class LambdaDestination implements logs.ILogSubscriptionDestination {
-  constructor(private readonly fn: lambda.IFunction) {
+  constructor(private readonly fn: lambda.IFunction, private readonly addPermissions: boolean=true) {
```
In all CDK constructs and also integration classes we put these kind of optional configuration arguments in a struct.

That way, when you're specifying it it looks like:

```ts
new LambdaDestination(lambda, { addPermissions: false })
```

Instead of:

```ts
new LambdaDestination(lambda, false)
```

Which is from first glance much easier to see what's going on.

Also, an options struct has the advantage over positional arugments that if we want to add more flags later on, it doesn't end up looking like:

```ts
new LambdaDestination(lambda, false, false, false, true, false, true);
```

zARODz11z commented:
@rix0rrr  I placed the options in a struct, I couldn't find an example from existing aws-cdk constructs, so let me know if I should modify it somehow. Also the linter is complaining because there hasn't been a change to a readme file.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 6d022cb8c5373b583ea4b8d5a38ba1f6eb2d1b4c
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1f595e8d-940c-4c30-bbf9-bc9594a3aad2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: cf9b14cb14632fa9969d470020ad0de3e5f281e2
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f0fba1e2-e0ca-4720-96a2-ad594089f2e8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

rix0rrr commented on packages/@aws-cdk/aws-logs-destinations/lib/lambda.ts:
```diff
@@ -4,22 +4,54 @@ import * as logs from '@aws-cdk/aws-logs';
+  /** Whether or not to add Lambda Permissions.
+   * @default true
+   */
+  readonly addPermissions?: boolean | undefined;
```
The `| undefined` is redundant with the `?`.

rix0rrr commented on packages/@aws-cdk/aws-logs-destinations/lib/lambda.ts:
```diff
@@ -4,22 +4,54 @@ import * as logs from '@aws-cdk/aws-logs';
+  readonly addPermissions?: boolean | undefined;
+}
+
+const defaultLambdaDestinationOptions: LambdaDestinationOptions = {
```
We generally don't do this, but interpret the absence of a value as the advertised default value.

That could take the forms of:

```ts
if (options.addPermissions !== false) {
  // undefined doesn't count as false, which means it gets interpreted the same way as true
}

// or

if (options.addPermissions ?? true) {
  // explicitly switch undefined to true, then compare
}
```

rix0rrr commented on packages/@aws-cdk/aws-logs-destinations/lib/lambda.ts:
```diff
@@ -4,22 +4,54 @@ import * as logs from '@aws-cdk/aws-logs';
-  constructor(private readonly fn: lambda.IFunction) {
+  /**  LambdaDestinationOptions */
+  options: LambdaDestinationOptions;
+  constructor(private readonly fn: lambda.IFunction, options?: LambdaDestinationOptions) {
```
```suggestion
  constructor(private readonly fn: lambda.IFunction, private readonly options: LambdaDestinationOptions = {}) {
```

rix0rrr commented on packages/@aws-cdk/aws-logs-destinations/lib/lambda.ts:
```diff
@@ -4,22 +4,54 @@ import * as logs from '@aws-cdk/aws-logs';
     return { arn: this.fn.functionArn };
   }
 }
+function setLambdaDestinationOptions(options: LambdaDestinationOptions | undefined) {
```
Don't need this if you're going to evaluate the default at read time.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: c102da50dbeb407d77d69128f109743bbee5d0cf
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=47161e6f-6a13-49c8-8dc7-595a7293f443%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 629086092a14552f1ce5f184542fb97af5a9156b
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=784033f8-9341-41cb-aeca-14a4bae4ac23%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

zARODz11z commented:
@rix0rrr I addressed your comments however I guess I need approval for some of the github workflows to run. I guess its a new feature from github to combat crypto farming.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 8df7cc88204befbaaf1a8780af4d3bf275fe132f
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=66dbd65f-be80-490b-8ce6-8ffdb8b99da3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 7d3794cb58a995e2474a55ef436809821afc46aa
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fb40d268-5917-4d64-9709-1aeff3f0a4a6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: fcd758c01cca17559e418b9fe35dc83af98cea7d
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ea5dc8d0-5ea6-4478-aa09-f7e1c1302b92%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

daisuke-yoshimoto commented:
@zARODz11z @rix0rrr 

Is there any update on this?
We are very troubled with the same problem. Would you please release this fix?

zARODz11z commented:
The branch of out of date with the main/master branch and now it needs another approval

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 02ae56e7f570d16dfae9558ac1661c97ce2eed54
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=32b512d1-ff23-47bc-b292-69541f1300ac%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 4c3dd5a383abcfa9edf9a6d8a62540cd3fe20edc
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4aed1a09-8573-4e49-a0a8-a3e514939c01%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: ceb09d29b6f429ba40851193993f526ac824e8b1
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2662c760-ed5a-494f-a594-cc4e98104da0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
